lib/src/css_mirroring/util_functions.dart

```diff
@@ -0,0 +1,109 @@
+part of scissors.src.css_mirroring.bidi_css_generator;
+
+/// Makes transaction from input string.
+TextEditTransaction _makeTransaction(String inputCss, String url) =>
+    new TextEditTransaction(inputCss, new SourceFile(inputCss, url: url));
+
+///  Checks if a topLevel tree node is direction independent.
+bool _areDirIndependentDirectives(TreeNode a, TreeNode b) =>
+    (a.runtimeType == b.runtimeType &&
+        (a is CharsetDirective ||
+            a is FontFaceDirective ||
+            a is ImportDirective ||
+            a is NamespaceDirective));
+
+bool _areRuleSets(TreeNode a, TreeNode b) => (a is RuleSet && b is RuleSet);
+
+bool _areDirDependentDirectives(TreeNode a, TreeNode b) =>
+    (a.runtimeType == b.runtimeType &&
+        (a is MediaDirective || a is HostDirective));
+
+bool _areDeclarations(TreeNode a, TreeNode b) =>
+    a is Declaration && b is Declaration;
+
+bool _areDeclarationsTextEqual(Declaration a, Declaration b) =>
+    a.span.text == b.span.text;
+
+
+/// A rule can be removed if all the declarations in the rule can be removed.
+bool _isRuleRemovable(_PendingRemovals removals, RuleSet rule) =>
+    removals._removalStartEndLocations.length ==
+        rule.declarationGroup.declarations.length;
+
+/// Checks if the declaration has to be removed based on the the Retention mode.
+bool _shouldRemoveDecl(
+    _RetentionMode mode, Declaration original, Declaration flipped) {
+  final bool isEqual = _areDeclarationsTextEqual(original, flipped);
+  return mode == _RetentionMode.keepBidiNeutral ? !isEqual : isEqual;
+}
+
+
+int _getNodeStart(TreeNode node) {
+  if (node is RuleSet) return node.span.start.offset;
+// In case of Directives since the node span start does not include '@'
+// so additional -1 is required.
+  return node.span.start.offset - 1;
+}
+
+int _getRuleSetEnd(List<RuleSet> ruleSets, int iTopLevel, int parentEnd) {
```

Comment:
iTopLevel doesn't mean much locally, maybe `ruleSetIndex` would be more appropriate?
(and please document parentEnd)